From 925f18848e0bb1f0d0fbf32d7a7cef76fce8bd90 Mon Sep 17 00:00:00 2001
From: Sumedh Sen <sumedhsen@google.com>
Date: Wed, 5 Mar 2025 14:42:42 -0800
Subject: [PATCH 4/7] Remove invalid null check for callingPackage

getLaunchedFromPackage() will never return null. As such, remove the null check while processing a session based install

Bug: 399881067
Bug: 364604008
Test: atest CtsPackageInstallTestCases CtsAppOps2TestCases
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:d733423089a4006c2cdd2357be6cf176e84ea150)
Merged-In: I4dc5fc8108c1ffffde82a5d774d4eb3805fbf0a3
Change-Id: I4dc5fc8108c1ffffde82a5d774d4eb3805fbf0a3
---
 .../packageinstaller/InstallStart.java        | 19 ++++++++++++-------
 1 file changed, 12 insertions(+), 7 deletions(-)

diff --git a/packages/PackageInstaller/src/com/android/packageinstaller/InstallStart.java b/packages/PackageInstaller/src/com/android/packageinstaller/InstallStart.java
index 0875a5cc277c..762a6a3811d4 100644
--- a/packages/PackageInstaller/src/com/android/packageinstaller/InstallStart.java
+++ b/packages/PackageInstaller/src/com/android/packageinstaller/InstallStart.java
@@ -29,6 +29,7 @@ import android.content.Intent;
 import android.content.pm.ApplicationInfo;
 import android.content.pm.IPackageManager;
 import android.content.pm.PackageInstaller;
+import android.content.pm.PackageInstaller.SessionInfo;
 import android.content.pm.PackageManager;
 import android.content.pm.ProviderInfo;
 import android.net.Uri;
@@ -45,12 +46,14 @@ public class InstallStart extends Activity {
     private static final String LOG_TAG = InstallStart.class.getSimpleName();
 
     private static final String DOWNLOADS_AUTHORITY = "downloads";
+    private PackageManager mPackageManager;
     private IPackageManager mIPackageManager;
     private boolean mAbortInstall = false;
 
     @Override
     protected void onCreate(@Nullable Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+        mPackageManager = getPackageManager();
         mIPackageManager = AppGlobals.getPackageManager();
         Intent intent = getIntent();
         String callingPackage;
@@ -66,12 +69,14 @@ public class InstallStart extends Activity {
         // If the activity was started via a PackageInstaller session, we retrieve the calling
         // package from that session
         final int sessionId = (isSessionInstall
-                ? intent.getIntExtra(PackageInstaller.EXTRA_SESSION_ID, -1)
-                : -1);
-        if (callingPackage == null && sessionId != -1) {
-            PackageInstaller packageInstaller = getPackageManager().getPackageInstaller();
+                ? intent.getIntExtra(PackageInstaller.EXTRA_SESSION_ID, SessionInfo.INVALID_ID)
+                : SessionInfo.INVALID_ID);
+        if (sessionId != SessionInfo.INVALID_ID) {
+            PackageInstaller packageInstaller = mPackageManager.getPackageInstaller();
             PackageInstaller.SessionInfo sessionInfo = packageInstaller.getSessionInfo(sessionId);
-            callingPackage = (sessionInfo != null) ? sessionInfo.getInstallerPackageName() : null;
+            if (sessionInfo != null) {
+                callingPackage = sessionInfo.getInstallerPackageName();
+            }
         }
 
         final ApplicationInfo sourceInfo = getSourceInfo(callingPackage);
@@ -168,7 +173,7 @@ public class InstallStart extends Activity {
     private ApplicationInfo getSourceInfo(@Nullable String callingPackage) {
         if (callingPackage != null) {
             try {
-                return getPackageManager().getApplicationInfo(callingPackage, 0);
+                return mPackageManager.getApplicationInfo(callingPackage, 0);
             } catch (PackageManager.NameNotFoundException ex) {
                 // ignore
             }
@@ -219,7 +224,7 @@ public class InstallStart extends Activity {
     }
 
     private boolean isSystemDownloadsProvider(int uid) {
-        final ProviderInfo downloadProviderPackage = getPackageManager().resolveContentProvider(
+        final ProviderInfo downloadProviderPackage = mPackageManager.resolveContentProvider(
                 DOWNLOADS_AUTHORITY, 0);
         if (downloadProviderPackage == null) {
             // There seems to be no currently enabled downloads provider on the system.
-- 
2.40.1

